### PR TITLE
Editable review

### PR DESCRIPTION
**Pull request by benedikthth**

### CODE REVIEW

File: src/api/apiFunctions.js [hunk 1]
```diff
@@ -0,0 +1,8 @@
+let handleErrors = (res) => {
```
Comment (Epli) [inline src/api/apiFunctions.js:1]: let handleError = (res) => res.ok ? res : throw Error(res.status)

Comment (Epli) [reply src/api/apiFunctions.js:1]: apiFunctions => errorHandler

File: src/pages/Login/Login.js [hunk 2]
```diff
@@ -48,11 +52,22 @@ function SignIn(props) {
...
-		else {
-			props.logIn(response);
-			props.history.push('/');
-		}
+		props.logIn(response);
+		props.history.push('/');
+	}).catch(x=>{
+		console.log(x);
+		
+		switch (Number(x.message)) {
```
Comment (Epli) [inline src/pages/Login/Login.js:60]: This should be handled on server side

File: src/pages/Login/Login.js [hunk 3]
```diff
@@ -8,6 +8,10 @@ import { makeStyles } from '@material-ui/core/styles';
 import { logIn } from 'actions';
 import Authentication from 'api/authentication';
 
+
+const LoginErrorMessage = 'Credentials did not match an account on our System';
```
Comment (Epli) [inline src/pages/Login/Login.js:12]: This should be handled on server side

File: src/pages/Profile/Profile.js [hunk 4]
```diff
@@ -42,9 +47,17 @@ const Profile = props => {
         setReviews(nreviews)
     }).bind(null, setReviews)
 
+
     let averageUpvotes = Math.round(reviews.map(x=> x.positiveVotes/(x.positiveVotes + x.negativeVotes)).reduce((a,b)=>a+b, 0)/reviews.length * 100 * 2)/2
 
-    console.log(averageUpvotes);
+
+    const handleClick = (event) => {
```
Comment (Epli) [inline src/pages/Profile/Profile.js:54]: const handleClick = (event) => setAnchorEl(event.currentTarget);

Comment (Epli) [reply src/pages/Profile/Profile.js:54]: const handleClose = () =>setAnchorEl(null);

File: src/pages/Teacher/EditableReview.js [hunk 5]
```diff
@@ -0,0 +1,173 @@
+
```
Comment (Epli) [inline src/pages/Teacher/EditableReview.js:1]: Remove commented code

File: src/pages/Teacher/EditableReview.js [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+}));
+
+const placeVote = (id, vote) => {
+    ReviewApi.placeVote(id, vote);
+
+    return vote;
+}
+
+
+function EditableReview(props) {
```
Comment (Epli) [inline src/pages/Teacher/EditableReview.js:81]: Why is this not just a Review? Looks like a status in Review in stead of a whole new component

File: src/pages/Teacher/Review.js [hunk 7]
```diff
@@ -1,11 +1,16 @@
...
 import { makeStyles } from '@material-ui/core/styles';
 import {ThumbUp, ThumbDown } from '@material-ui/icons'
 import PieChart from 'components/Piechart';
 import debounce from 'util/debounce';
 import { connect } from 'react-redux';
-import ReviewApi from '../../api/reviews'
+import ReviewApi from '../../api/reviews';
+import Question from './ReviewQuestion';
+import {ToggleButton } from '@material-ui/lab';
+import EditIcon from '@material-ui/icons/Edit';
```
Comment (Epli) [inline src/pages/Teacher/Review.js:11]: Use the icons import above

File: src/pages/Teacher/ReviewQuestion.js [hunk 8]
```diff
@@ -11,7 +11,9 @@ const labels = {
     4: 'Excellent',
     5: 'Awesome'
 };
-  
+
```
Comment (Epli) [inline src/pages/Teacher/ReviewQuestion.js:14]: Remove the newlines

File: src/pages/Teacher/Review.js [hunk 9]
```diff
@@ -77,11 +82,33 @@ const placeVote = (id, vote) => {
 
 function Review(props) {
     const classes = useStyles()
-    const [toggle, setToggle] = useState(false);
     const review = props.review;
```
Comment (Epli) [inline src/pages/Teacher/Review.js:85]: Read props in oneline, example: const { review, canEdit } = props;

File: src/pages/Teacher/Review.js [hunk 10]
```diff
@@ -77,11 +82,33 @@ const placeVote = (id, vote) => {
...
+    const [lectureQuality, setLectureQuality] = useState(review.lectureQuality)    
+    const [wouldTakeAgainQuality, setWouldTakeAgainQuality] = useState(review.wouldTakeAgainQuality)    
+    const [text, setText] = useState(review.text)
+    
+
+    const handleTextChange = (ev) => {
+        setText(ev.target.value)
+    }
+
+    const editReview = (id, teacherQuality, text, lectureQuality, wouldTakeAgainQuality, teacherID) => {
```
Comment (Epli) [inline src/pages/Teacher/Review.js:101]: pass in reivew instead of 5 params

File: src/pages/Teacher/Review.js [hunk 11]
```diff
@@ -77,11 +82,33 @@ const placeVote = (id, vote) => {
...
+    const [editing, setEditing] = useState(false);
+    const [toggle, setToggle] = useState(false);
+
+    const [teacherQuality, setTeacherQuality] = useState(review.teacherQuality)    
+    const [lectureQuality, setLectureQuality] = useState(review.lectureQuality)    
+    const [wouldTakeAgainQuality, setWouldTakeAgainQuality] = useState(review.wouldTakeAgainQuality)    
+    const [text, setText] = useState(review.text)
+    
+
+    const handleTextChange = (ev) => {
```
Comment (Epli) [inline src/pages/Teacher/Review.js:97]: online

File: src/pages/Teacher/Review.js [hunk 12]
```diff
@@ -77,11 +82,33 @@ const placeVote = (id, vote) => {
...
+        setText(ev.target.value)
+    }
+
+    const editReview = (id, teacherQuality, text, lectureQuality, wouldTakeAgainQuality, teacherID) => {
+        let review = {teacherID, text, teacherQuality, lectureQuality, wouldTakeAgainQuality}
+        
+        ReviewApi.EditReview(review).then(x=>{
```
Comment (Epli) [inline src/pages/Teacher/Review.js:104]: online

File: src/pages/Teacher/Review.js [hunk 13]
```diff
@@ -77,11 +82,33 @@ const placeVote = (id, vote) => {
...
+            setEditing(false);
+        })
+    }
+
     console.log(review);
```
Comment (Epli) [inline src/pages/Teacher/Review.js:109]: Remove

File: src/pages/Teacher/Review.js [hunk 14]
```diff
@@ -77,11 +82,33 @@ const placeVote = (id, vote) => {
...
+    const canEdit = props.canEdit;
     const [vote, setVote] = useState(review.vote);
+    const [editing, setEditing] = useState(false);
+    const [toggle, setToggle] = useState(false);
+
+    const [teacherQuality, setTeacherQuality] = useState(review.teacherQuality)    
+    const [lectureQuality, setLectureQuality] = useState(review.lectureQuality)    
+    const [wouldTakeAgainQuality, setWouldTakeAgainQuality] = useState(review.wouldTakeAgainQuality)    
+    const [text, setText] = useState(review.text)
+
```
Comment (Epli) [inline src/pages/Teacher/Review.js:95]: Whiteline

File: src/pages/Teacher/Review.js [hunk 15]
```diff
@@ -127,17 +179,46 @@ function Review(props) {
                 </ExpansionPanelSummary>
                 <ExpansionPanelDetails>
                     <Grid container className={classes.summary}>
+
+
                         <Grid item xs={4}>
-                            {toggle && <PieChart title="Lecture Quality" isAnimationActive={toggle} maxValue={5} value={review.lectureQuality} width="100%" height={250}/>}
+                            {toggle && <>
```
Comment (Epli) [inline src/pages/Teacher/Review.js:185]: Chagne to: toggle && editing

File: src/pages/Teacher/Review.js [hunk 16]
```diff
@@ -96,18 +123,43 @@ function Review(props) {
...
+                                        >
+                                        <EditIcon />
+                                    </ToggleButton>
+                                }
+                                </Grid>
+                            </Grid>
+
+                            <Grid container>
+                                
+                                { editing ? <TextField value={text} onChange={handleTextChange}></TextField>:
```
Comment (Epli) [inline src/pages/Teacher/Review.js:150]: Change to: editing &&

File: src/pages/Teacher/Review.js [hunk 17]
```diff
@@ -96,18 +123,43 @@ function Review(props) {
...
+                                    <Typography component="p" variant="body2" className={classes.commentHeader}>
+                                        {review.userName} - {new Date(review.created).toLocaleDateString()} {new Date(review.created).toLocaleTimeString()}
+                                    </Typography>
+                                </Grid>
+                                <Grid item xs={4}>        
+                                {canEdit &&
+                                    <ToggleButton
+                                        value="check"
+                                        selected={editing}
+                                        onChange={(e) => {
```
Comment (Epli) [inline src/pages/Teacher/Review.js:137]: Is e.stopPropagation() nessary?

File: src/pages/Teacher/Review.js [hunk 18]
```diff
@@ -127,17 +179,46 @@ function Review(props) {
...
+                                <PieChart title="Would Take Again Quality"  isAnimationActive={toggle} maxValue={5} value={wouldTakeAgainQuality} width="100%" height={250}/>
+                                }</>} 
+                                    
                         </Grid>
+
+                        
                     </Grid>
+
+                    {editing && <Button variant="contained" 
+                        onClick={editReview.bind(null,
```
Comment (Epli) [inline src/pages/Teacher/Review.js:212]: Arrow function over bind, ( bind is old school )

File: src/pages/Teacher/Teacher.js [hunk 19]
```diff
@@ -12,7 +12,7 @@ import ReviewForm from './ReviewForm'
 import HistogramCard from 'components/HistogramCard'
 import { connect } from 'react-redux';
 import ReviewApi from 'api/reviews'
-
+// import EditableReview from './EditableReview'
```
Comment (Epli) [inline src/pages/Teacher/Teacher.js:15]: Remove

File: src/pages/Teacher/Teacher.js [hunk 20]
```diff
@@ -121,14 +121,15 @@ function Teacher (props) {
         setReviews(nreviews)
     }).bind(null, setReviews)
 
+    let currentUserHasReviewed = loggedIn;
```
Comment (Epli) [inline src/pages/Teacher/Teacher.js:124]: This can be done by only using one variable

File: src/pages/Teacher/Teacher.js [hunk 21]
```diff
@@ -212,7 +213,11 @@ function Teacher (props) {
             }
 
             <ReviewContainer sort={setOrderby}>
-                {(reviews.sort((a, b) => order(a, b, orderBy))).map((rev, ind)=> <Review key={ind} removeReview={removeReview} review={rev}/>)}
+                {(reviews.sort((a, b) => order(a, b, orderBy))).map((rev, ind)=> 
+                // (rev.userId !== props.userStore.user.id)?
```
Comment (Epli) [inline src/pages/Teacher/Teacher.js:217]: Remove Comments

File: src/pages/Teacher/Teacher.js [hunk 22]
```diff
@@ -212,7 +213,11 @@ function Teacher (props) {
             }
 
             <ReviewContainer sort={setOrderby}>
-                {(reviews.sort((a, b) => order(a, b, orderBy))).map((rev, ind)=> <Review key={ind} removeReview={removeReview} review={rev}/>)}
+                {(reviews.sort((a, b) => order(a, b, orderBy))).map((rev, ind)=> 
+                // (rev.userId !== props.userStore.user.id)? 
+                <Review key={ind} removeReview={removeReview} canEdit={loggedIn?rev.userId === userStore.user.id:false} review={rev}/>
```
Comment (Epli) [inline src/pages/Teacher/Teacher.js:218]: Change to: loggedIn && rev.userId === userStore.user.id

Comment (Epli) [reply src/pages/Teacher/Teacher.js:218]: Rename ind to index

File: src/pages/Login/Login.js (hunk 2 above)
Comment (benedikthth) [reply src/pages/Login/Login.js:60]: this is handled on the server side. the presentation is decided on the client. Everything else can fail quietly

File: src/pages/Login/Login.js (hunk 3 above)
Comment (benedikthth) [reply src/pages/Login/Login.js:12]: It is. The reason we do this here is when we do localization we want all errors to be presented clientside.

File: src/pages/Teacher/EditableReview.js (hunk 6 above)
Comment (benedikthth) [reply src/pages/Teacher/EditableReview.js:81]: removing this

File: src/pages/Teacher/Review.js (hunk 16 above)
Comment (benedikthth) [reply src/pages/Teacher/Review.js:150]: why not have a ternary operator when we have this kind of conditional operation?

File: src/pages/Teacher/Review.js (hunk 17 above)
Comment (benedikthth) [reply src/pages/Teacher/Review.js:137]: yes, otherwise the panel will toggle every time we press the edit button

**Pull request merged**